crypto/engine/eng_int.h

```diff
@@ -151,6 +168,11 @@ struct engine_st {
     ENGINE_SSL_CLIENT_CERT_PTR load_ssl_client_cert;
     const ENGINE_CMD_DEFN *cmd_defns;
     int flags;
+    /*
+     * The access lock for this engine, if both this and the global lock
+     * are required, the global lock *must* be taken first.
+     */
```

Comment:
If I may, I think this comment is a bit vague.  Want to make a rule that we can follow, then it should say that the global lock should only be used with global things (such as the libcrypto internal list of engines) while the ENGINE lock should be used when accessing ENGINE internals.

Reply:
I'll try to rework this comment in the morning.  Unfortunately, I couldn't see how to do the functional reference count, which is an engine thing, using the per engine lock.

Look at `engine_unlocked_init`.  The problem is the test for the functional reference count being zero meaning the init function should be called clashing with the increment of the functional reference count later.

As it stands, multiple threads could check for the zero before any of them increments it.  There is no guarantee that the init function is idempotent, so things can fail when it is called several times.  Rearranging the code, to increment the count when reading it also fails.  If the counter is zero, and the init function returns false, the counter has to be decremented again but if another thread jumps in it is possible for neither to run the finish function.  These two situations pretty much rule out using atomic operations.  I'm hesitant to have the init call so it runs under the per engine lock in case it does something that requires this lock.

It would be great if the functional reference count could be made atomic or locked using the per engine lock but I don't see a nice way to do it (at least not yet).

Reply:
The under global lock comment applies here, there is no chance the value can change, so it can be done.  It still requires a read operation on the counter for elsewhere which means it cannot be a `CRYPTO_REF_COUNT` counter.